docs/design/backup_and_restore.md

```diff
@@ -0,0 +1,202 @@
+    * Destination
+    * Oldest Data
+        * Version
+        * Timestamp
+    * Latest Restorable Point
+        * Version
+        * Timestamp
+
+In the output for `kubectl get`, we will include the name of the backup, the generation, whether it is fully reconciled, the oldest data timestamp, and the latest restorable timestamp. For this purpose a backup will be considered fully reconciled if it has no conditions in its status.
+
+The possible conditions on the backup will be:
+
+* IncorrectDeployment
+* DeploymentRollingOut
+* IncorrectBackupState
+* MisconfiguredBackup
+
+Wherever we have timestamp formats in the resource specs, they will be interpreted as [Go time formats](https://pkg.go.dev/time#Time.Format).
+
+### Restore Resource
+
+We will create a FoundationDBRestore resource with the following spec:
+
+* Backup URL
+* Destination cluster name
+* Desired state (running, stopped) (optional, defaults to running)
+* Key range to restore (optional, defaults to the full keyspace)
+* Version to restore (optional, defaults to latest version)
+* Timestamp to restore (optional, defaults to latest version)
+* Source cluster name (optional, required if using restore timestamps)
+* Restore tag (optional)
+
+All fields in a restore are immutable.
+
+If the restore version and restore timestamp are both specified, the version will take precedence.
+
+### Reconciling Backups
+
+The backup reconciler will run the following stages:
+
+* Check backup status
+* Configure backup agents
+* Toggle backup running
+* Modify backup
+* Toggle backup paused
+* Check backup status
+
+We will use the `fdbbackup` tool to modify the backup state. We will use the backup status that is found within the database status to determine the current state of the backup.
+
+#### Check Backup Status
+
+In the `Check Backup Status` stage, we will compare the running state and validate the following things:
+
+* The state of the backup matches the configured state in the spec
+* If the backup is configured to be running, it has completed an initial snapshot
+* The backup parameters match the parameters in the spec
+* The backup deployment exists, matches the deployment configuration for the spec, and is fully deployed
+
+If any of these checks fail, the backup will be marked as unreconciled, but reconciliation will continue through the later stages whether it is reconciled or not.
+
+#### Configure Backup Agents
+
+In the `Configure Backup Agents` stage, we will create a backup deployment that runs FDB's `backup_agent` process. The agents will be configured through a deployment object in the cluster spec. The deployment will have the following defaults:
+
+* The deployment will have the same name as the backup resource
+* The deployment will have 3 replicas
+* The deployment will have an init container running the `foundationdb/foundationdb-kubernetes-sidecar` image, configured to copy the cluster file from the same config map that the operator creates for the cluster.
```

Comment:
The limitation of this is that we can't use this mechanism to backup/restore clusters in a different namespace/cluster since the `ConfigMap` can only be consumed as a volume in the same namespace. I think for the initial steps that's fine but I could think of a setup where the backup agents are running in a different namespace to ensure the backups are not deleted with the clusters if someone deletes the namespace (intended or not).

*edit* I see you mention that below. Are we planning to support that in the future or should we just assume this as a limitation of the design?

Reply:
I'm happy to accept it as a limitation of the design. If we want to keep backup data as a safety mechanism after a namespace is deleted, we can look into expressing that as a wait time in the finalizer logic.